### PR TITLE
Bring back sorting when needed to fix PlayTo

### LINKED ISSUE #1724: "Play to" provides no way to order files
**Describe the bug**
When using the 'Play to' feature, there's currently no way for the receiving client to know what order files should be played in.

**To Reproduce**
1. Be logged into 2 clients that support the 'Play to' (web client, android app, kodi)
2. Send a playlist/collection from one device to another
4. Playlist/collection contents play in a seemingly random order

**Expected behavior**
Items should play in the same order as displayed

**Logs**
Mixed logs and findings from investigating
https://pastebin.com/8zct2g76

**System (please complete the following information):**
 - OS: Docker
 - Browser: Firefox
 - Jellyfin Version: 10.3.7
 - Reverse proxy: nginx

**Additional context**
Originally discovered while investigating https://github.com/jellyfin/jellyfin-kodi/issues/49, but this also occurs with the mobile client or 2 different web clients.  It seems like originally this was handled by the server sorting item IDs before sending instead of by the clients, but was removed in https://github.com/jellyfin/jellyfin/pull/1011.  Currently, the server only sends a list of ItemIds of the contents of what is being sent (Playlist, Collection, etc).  There's no way for the clients to know what order these items should be played in without more information about the parent.  It seems like either the server needs to sort items before sending them, or it needs to expose the ItemID of what's being sent (Playlist, Collection, etc) so the clients can look up that data themselves.

### PR DESCRIPTION
**Pull request by JustAMan**
**Changes**

This is partial revert of https://github.com/jellyfin/jellyfin/pull/1011 with minor improvement (don't sort an array of 1 item).

Note: it was originally stated #1011 sped up loading of interface tremendously. I cannot find where this particular method would be called except in "PlayTo" functionality, so here's hoping nothing breaks 🤞 

I've also implemented faster method of sorting.

**Issues**

Fixes #1724

**Notes**
This requires proper testing.

### CODE REVIEW

**Comment by mcarlton00** [comment]
I'm not seeing any performance issues during page load or during sending content to a remote player.

Sending from the Android app or web browser to other browsers fully works with no ordering issues.  Tested seasons of shows, music albums, and music playlists.

Music playback to Mopidy sends in the proper order, but the wrong song starts playing (I don't think the API tells us what element in the list to start playing at)

When sending to Kodi, TV Shows work as expected.  Tested both individually and seasons and things seem to consistently come through as expected.  Sending a music playlist or album works as expected.  Kodi receives them in the correct order and begins playing at the first track.  Choosing a specific element of a playlist (music or videos)t or album also just begins playing the first item.

I need to check the API and verify if I'm receiving a method to know which element in the list I should be starting at, but judging by my findings when investigating #1724 initially, I don't think there's any indication where to start.

Payload from previous findings for reference:
```
{'ControllingUserId': '3920b99003194448a2a9d4f0bedd3d82',
 'Id': '21b4f3891c98bca31fe0ca9dd85f697a',
 'ItemIds': ['df3e113a990a813a434e07c0f7e3d7fa',
             '656e10dc916933ff598a5b68babde709',
             'ee80cac1f69a72999f96b502c5be9639'],
 'PlayCommand': 'PlayNow'}
```

**Comment by mcarlton00** [comment]
Just checked on the API and it's already returning the data needed about where in the list to start playback, Kodi and Mopidy just aren't utilizing it properly yet.

```python
(Pdb) pp message
'Play'
(Pdb) pp data
{'ControllingUserId': '3920b99003194448a2a9d4f0bedd3d82',
 'Id': '64ca70378ea31b4ea7fa4758b3f3fb00',
 'ItemIds': ['d9ab6c5b9e2d4322792117e6b69527be',
             '73af9c5aaf113f5f5301792351855870',
             '00049cb6463f62267eae84d2feb4e25b',
             '332870fbca08d028bde9a8b725832b04',
             'c29e275654119cd795ab6b3ea8de6107',
             '666c1e818c1e836315f5fa543155bcf8'],
 'PlayCommand': 'PlayNow',
 'StartIndex': 3}
```

File: MediaBrowser.Controller/Entities/Folder.cs [hunk 1]
```diff
@@ -807,11 +807,42 @@ private bool RequiresPostFiltering(InternalItemsQuery query)
             return false;
         }
 
+        private IReadOnlyList<BaseItem> SortItemsByRequest(InternalItemsQuery query, IReadOnlyList<BaseItem> items)
```
Comment (Bond-009) [inline MediaBrowser.Controller/Entities/Folder.cs]:
```suggestion
        private static BaseItem[] SortItemsByRequest(InternalItemsQuery query, IReadOnlyList<BaseItem> items)
```

File: MediaBrowser.Controller/Entities/Folder.cs [hunk 2]
```diff
@@ -807,11 +807,42 @@ private bool RequiresPostFiltering(InternalItemsQuery query)
...
+                {
+                    index++;
+                }
+            }
+
+            var newItems = new BaseItem[size];
+            foreach(var item in items)
+            {
+                newItems[positions[item.Id]] = item;
+            }
```
Comment (Bond-009) [inline MediaBrowser.Controller/Entities/Folder.cs]:
```suggestion
            }

```

File: MediaBrowser.Controller/Entities/Folder.cs [hunk 3]
```diff
@@ -807,11 +807,42 @@ private bool RequiresPostFiltering(InternalItemsQuery query)
...
+            for (int i = 0; i < ids.Length; i++)
+            {
+                if (positions.TryAdd(ids[i], index))
+                {
+                    index++;
+                }
+            }
+
+            var newItems = new BaseItem[size];
+            foreach(var item in items)
```
Comment (Bond-009) [inline MediaBrowser.Controller/Entities/Folder.cs]:
```suggestion
            foreach (var item in items)
```
If you want ever better perf, change this into a `for` loop

File: MediaBrowser.Controller/Entities/Folder.cs [hunk 4]
```diff
@@ -807,11 +807,42 @@ private bool RequiresPostFiltering(InternalItemsQuery query)
...
         {
             if (query.ItemIds.Length > 0)
             {
-                return LibraryManager.GetItemsResult(query);
+                var result = LibraryManager.GetItemsResult(query);
+
+                if (query.OrderBy.Count == 0 && query.ItemIds.Length > 1)
+                {
+                    result.Items = SortItemsByRequest(query, result.Items);
+                }
```
Comment (Bond-009) [inline MediaBrowser.Controller/Entities/Folder.cs]:
```suggestion
                }

```

File: MediaBrowser.Controller/Entities/Folder.cs [hunk 5]
```diff
@@ -823,7 +854,13 @@ public IReadOnlyList<BaseItem> GetItemList(InternalItemsQuery query)
 
             if (query.ItemIds.Length > 0)
             {
-                return LibraryManager.GetItemList(query);
+                var result = LibraryManager.GetItemList(query);
+
+                if (query.OrderBy.Count == 0 && query.ItemIds.Length > 1)
+                {
+                    return SortItemsByRequest(query, result);
+                }
```
Comment (Bond-009) [inline MediaBrowser.Controller/Entities/Folder.cs]:
```suggestion
                }

```

File: MediaBrowser.Controller/Entities/Folder.cs (hunk 3 above)
Comment (JustAMan) [reply MediaBrowser.Controller/Entities/Folder.cs]: I've replaced `foreach` with `for`

**Comment by mcarlton00** [comment]
Tested updates, still working fine for other Web sessions and Mopidy.  Seems to be a bit faster than before when you trigger a large number at once (triggered 200 item playlist and it got processed and started playing in under ~3 seconds to Mopidy.  It was even faster from Web -> Web)

**Pull request merged**